File: src/behav3d_cpp/src/target_builder.cpp
// =============================================================================
//   ____  _____ _   _    ___     _______ ____
//  | __ )| ____| | | |  / \ \   / /___ /|  _ \ 
//  |  _ \|  _| | |_| | / _ \ \ / /  |_ \| | | |
//  | |_) | |___|  _  |/ ___ \ V /  ___) | |_| |
//  |____/|_____|_| |_/_/   \_\_/  |____/|____/
//
// Author: Özgüç Bertuğ Çapunaman <ozca@iti.sdu.dk>
// Maintainers:
//   - Lucas José Helle <luh@iti.sdu.dk>
//   - Joseph Milad Wadie Naguib <jomi@iti.sdu.dk>
// Institute: University of Southern Denmark (Syddansk Universitet)
// Date: 2025-07
// =============================================================================

#include <Eigen/Dense>
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/LinearMath/Matrix3x3.h>

#include "behav3d_cpp/util.hpp"

namespace behav3d::target_builder
{

    using behav3d::util::deg2rad;
    using behav3d::util::fromRPY;
    using geometry_msgs::msg::PoseStamped;

    // Build a timestamped PoseStamped with explicit position & quaternion
    PoseStamped poseStamped(double x, double y, double z,
                            double qx, double qy, double qz, double qw,
                            const std::string &frame)
    {
        PoseStamped ps;
        ps.header.frame_id = frame;
        ps.header.stamp = rclcpp::Clock().now();
        ps.pose.position.x = x;
        ps.pose.position.y = y;
        ps.pose.position.z = z;
        ps.pose.orientation.x = qx;
        ps.pose.orientation.y = qy;
        ps.pose.orientation.z = qz;
        ps.pose.orientation.w = qw;
        return ps;
    }

    PoseStamped worldXY(double x, double y, double z,
                        const std::string &frame)
    {
        tf2::Quaternion q(0, 0, 0, 1);
        return poseStamped(x, y, z, q.x(), q.y(), q.z(), q.w(), frame);
    }

    PoseStamped worldXZ(double x, double z, double y,
                        const std::string &frame)
    {
        tf2::Quaternion q(0.7071067811865475, 0, 0, 0.7071067811865475);
        return poseStamped(x, y, z, q.x(), q.y(), q.z(), q.w(), frame);
    }

    PoseStamped worldYZ(double y, double z, double x,
                        const std::string &frame)
    {
        tf2::Quaternion q(0.5, 0.5, 0.5, 0.5);
        return poseStamped(x, y, z, q.x(), q.y(), q.z(), q.w(), frame);
    }

    // Convert PoseStamped → Eigen::Isometry3d  (pose → matrix)
    Eigen::Isometry3d toIso(const PoseStamped &p)
    {
        Eigen::Isometry3d iso = Eigen::Isometry3d::Identity();
        iso.translation() = Eigen::Vector3d(p.pose.position.x,
                                            p.pose.position.y,
                                            p.pose.position.z);
        Eigen::Quaterniond q(p.pose.orientation.w,
                             p.pose.orientation.x,
                             p.pose.orientation.y,
                             p.pose.orientation.z);
        iso.linear() = q.toRotationMatrix();
        return iso;
    }

    // Convert Eigen::Isometry3d → PoseStamped  (matrix → pose)
    PoseStamped fromIso(const Eigen::Isometry3d &iso,
                        const std::string &frame)
    {
        PoseStamped ps;
        ps.header.frame_id = frame;
        ps.header.stamp = rclcpp::Clock().now();
        ps.pose.position.x = iso.translation().x();
        ps.pose.position.y = iso.translation().y();
        ps.pose.position.z = iso.translation().z();

        Eigen::Quaterniond q(iso.rotation());
        ps.pose.orientation.x = q.x();
        ps.pose.orientation.y = q.y();
        ps.pose.orientation.z = q.z();
        ps.pose.orientation.w = q.w();
        return ps;
    }

    // Translate pose by vector d in its own frame
    PoseStamped translate(const PoseStamped &in,
                          const Eigen::Vector3d &d)
    {
        Eigen::Isometry3d iso = toIso(in);
        iso.translate(d);
        return fromIso(iso, in.header.frame_id);
    }

    // Rotate pose by RPY (degrees if `degrees==true`)
    PoseStamped rotateEuler(const PoseStamped &in,
                            const Eigen::Vector3d &rpy,
                            bool degrees)
    {
        Eigen::Isometry3d iso = toIso(in);
        iso.rotate(fromRPY(rpy, degrees));
        return fromIso(iso, in.header.frame_id);
    }

    // Apply relative translation & rotation to pose
    PoseStamped transformRel(const PoseStamped &in,
                             const Eigen::Vector3d &t,
                             const Eigen::Quaterniond &q)
    {
        Eigen::Isometry3d iso = toIso(in);
        iso.translate(t);
        iso.rotate(q);
        return fromIso(iso, in.header.frame_id);
    }

    // change-basis: embed a pose defined in a local frame into another pose's frame
    PoseStamped changeBasis(const PoseStamped &basis,
                            const PoseStamped &local)
    {
        Eigen::Isometry3d iso_basis = toIso(basis); // basis → world
        Eigen::Isometry3d iso_local = toIso(local); // pose  → basis
        Eigen::Isometry3d iso_world = iso_basis * iso_local;
        return fromIso(iso_world, basis.header.frame_id);
    }

    // rebase: transform a pose from one frame to another
    PoseStamped rebase(const PoseStamped &pose,
                       const PoseStamped &src,
                       const PoseStamped &dst)
    {
        // src.header.frame_id → world
        Eigen::Isometry3d iso_src = toIso(src);
        // pose in src frame
        Eigen::Isometry3d iso_pose = toIso(pose);
        // world pose
        Eigen::Isometry3d iso_world = iso_src * iso_pose;
        // dst.header.frame_id → world
        Eigen::Isometry3d iso_dst = toIso(dst);
        // express world in dst frame
        Eigen::Isometry3d iso_new = iso_dst.inverse() * iso_world;
        return fromIso(iso_new, dst.header.frame_id);
    }

    // axis helpers
    static Eigen::Vector3d axis(const PoseStamped &p, int idx)
    {
        tf2::Quaternion q(p.pose.orientation.x,
                          p.pose.orientation.y,
                          p.pose.orientation.z,
                          p.pose.orientation.w);
        tf2::Matrix3x3 m(q);
        tf2::Vector3 v = m.getColumn(idx);
        return {v.x(), v.y(), v.z()};
    }

    Eigen::Vector3d xAxis(const PoseStamped &p) { return axis(p, 0); }
    Eigen::Vector3d yAxis(const PoseStamped &p) { return axis(p, 1); }
    Eigen::Vector3d zAxis(const PoseStamped &p) { return axis(p, 2); }

    // Pure roll: align +X with desired axis, keep +Z fixed
    PoseStamped alignTarget(const PoseStamped &in,
                            const Eigen::Vector3d &desired_x)
    {
        // Keep current +Z so the operation is a *pure roll*
        Eigen::Isometry3d iso = toIso(in);

        const Eigen::Vector3d z = iso.linear().col(2).normalized();

        // Project desired X into the plane orthogonal to Z
        Eigen::Vector3d x_proj = desired_x - desired_x.dot(z) * z;

        // If projection is too small (desired_x ‖ Z) choose a fallback axis
        if (x_proj.squaredNorm() < 1e-10)
        {
            Eigen::Vector3d fallback = iso.linear().col(1); // camera +Y
            x_proj = fallback - fallback.dot(z) * z;
        }

        const Eigen::Vector3d x = x_proj.normalized();
        const Eigen::Vector3d y = z.cross(x).normalized(); // completes right‑handed basis

        iso.linear().col(0) = x;
        iso.linear().col(1) = y;
        iso.linear().col(2) = z; // unchanged

        return fromIso(iso, in.header.frame_id);
    }

    // Re-orient pose so +Z = new_normal, keep basis orthonormal
    PoseStamped adjustTarget(const PoseStamped &in,
                             const Eigen::Vector3d &new_normal)
    {
        Eigen::Isometry3d iso = toIso(in);

        const Eigen::Vector3d z = new_normal.normalized();
        Eigen::Vector3d x = iso.linear().col(0);

        // If old X is almost parallel to new Z, pick a safe world axis
        if (std::abs(x.dot(z)) > 0.95)
        {
            x = std::abs(z.dot(Eigen::Vector3d::UnitX())) < 0.8
                    ? Eigen::Vector3d::UnitX()
                    : Eigen::Vector3d::UnitY();
        }

        Eigen::Vector3d y = z.cross(x).normalized();
        x = y.cross(z); // ensure orthonormality

        iso.linear().col(0) = x;
        iso.linear().col(1) = y;
        iso.linear().col(2) = z;

        return fromIso(iso, in.header.frame_id);
    }

    // Return orientation as (axis, angle)
    std::pair<Eigen::Vector3d, double> axisAngle(const PoseStamped &p)
    {
        Eigen::Quaterniond q(p.pose.orientation.w,
                             p.pose.orientation.x,
                             p.pose.orientation.y,
                             p.pose.orientation.z);
        Eigen::AngleAxisd aa(q);
        return {aa.axis(), aa.angle()};
    }

    // Linear/SLERP blend between poses A and B (t∈[0,1])
    PoseStamped poseBetween(const PoseStamped &A,
                            const PoseStamped &B,
                            double t)
    {
        Eigen::Isometry3d a = toIso(A);
        Eigen::Isometry3d b = toIso(B);

        Eigen::Vector3d p = a.translation() * (1.0 - t) + b.translation() * t;
        Eigen::Quaterniond qa(a.rotation()), qb(b.rotation());
        Eigen::Quaterniond q = qa.slerp(t, qb);

        Eigen::Isometry3d iso = Eigen::Isometry3d::Identity();
        iso.linear() = q.toRotationMatrix();
        iso.translation() = p;

        return fromIso(iso, A.header.frame_id);
    }

    // Mirror pose across plane defined by point p0 and normal n
    PoseStamped mirrorAcrossPlane(const PoseStamped &pose,
                                  const Eigen::Vector3d &n,
                                  const Eigen::Vector3d &p0)
    {
        const Eigen::Vector3d n_n = n.normalized();

        Eigen::Vector3d pos(pose.pose.position.x,
                            pose.pose.position.y,
                            pose.pose.position.z);
        Eigen::Vector3d mirrored_pos =
            pos - 2.0 * (pos - p0).dot(n_n) * n_n;

        Eigen::Quaterniond q(pose.pose.orientation.w,
                             pose.pose.orientation.x,
                             pose.pose.orientation.y,
                             pose.pose.orientation.z);

        const Eigen::Quaterniond reflect(Eigen::AngleAxisd(M_PI, n_n));
        Eigen::Quaterniond mirrored_q = reflect * q;

        return poseStamped(mirrored_pos.x(), mirrored_pos.y(), mirrored_pos.z(),
                           mirrored_q.x(), mirrored_q.y(), mirrored_q.z(), mirrored_q.w(),
                           pose.header.frame_id);
    }

    // Average a vector of poses (mean position, hemiconstrained mean quaternion)
    PoseStamped average(const std::vector<PoseStamped> &poses)
    {
        if (poses.empty())
            return PoseStamped{};

        Eigen::Vector3d c = Eigen::Vector3d::Zero();
        Eigen::Quaterniond accum(0, 0, 0, 0);

        for (const auto &p : poses)
        {
            c += Eigen::Vector3d(p.pose.position.x,
                                 p.pose.position.y,
                                 p.pose.position.z);
            Eigen::Quaterniond q(p.pose.orientation.w,
                                 p.pose.orientation.x,
                                 p.pose.orientation.y,
                                 p.pose.orientation.z);
            if (q.w() < 0)
                q.coeffs() = -q.coeffs(); // hemiconstraint
            accum.coeffs() += q.coeffs();
        }

        c /= poses.size();
        accum.normalize();

        return poseStamped(c.x(), c.y(), c.z(),
                           accum.x(), accum.y(), accum.z(), accum.w(),
                           poses.front().header.frame_id);
    }

    /// Move the pose along its *local* +Z axis by `offset_dist` metres.
    PoseStamped offsetTarget(const PoseStamped &in,
                             double offset_dist)
    {
        // `translate` already interprets the vector in the pose’s own frame.
        return translate(in, Eigen::Vector3d(0.0, 0.0, offset_dist));
    }

    /// Replace the position (x, y, z) while preserving orientation.
    PoseStamped setTargetOrigin(const PoseStamped &in,
                                double x, double y, double z)
    {
        PoseStamped out = in; // copy orientation, header, etc.
        out.pose.position.x = x;
        out.pose.position.y = y;
        out.pose.position.z = z;
        out.header.stamp = rclcpp::Clock().now();
        return out;
    }

    /// Optionally flip the local X and/or Y axes; recompute +Z to keep a
    /// right‑handed, orthonormal basis.
    PoseStamped flipTargetAxes(const PoseStamped &in,
                               bool flip_x,
                               bool flip_y)
    {
        Eigen::Isometry3d iso = toIso(in);
        Eigen::Matrix3d R = iso.rotation();

        if (flip_x)
            R.col(0) *= -1.0;
        if (flip_y)
            R.col(1) *= -1.0;

        Eigen::Vector3d x = R.col(0).normalized();
        Eigen::Vector3d y = R.col(1).normalized();
        Eigen::Vector3d z = x.cross(y).normalized();

        // Ensure a right‑handed frame (determinant > 0)
        if (x.cross(y).dot(z) < 0.0)
            z *= -1.0;

        R.col(0) = x;
        R.col(1) = y;
        R.col(2) = z;
        iso.linear() = R;

        return fromIso(iso, in.header.frame_id);
    }

    /// Swap the local X and Y axes; adjust Z to restore orthonormality.
    PoseStamped swapTargetAxes(const PoseStamped &in)
    {
        Eigen::Isometry3d iso = toIso(in);
        Eigen::Matrix3d R = iso.rotation();

        R.col(0).swap(R.col(1)); // X ↔ Y

        Eigen::Vector3d x = R.col(0).normalized();
        Eigen::Vector3d y = R.col(1).normalized();
        Eigen::Vector3d z = x.cross(y).normalized();

        // Maintain right‑handed orientation.
        if (x.cross(y).dot(z) < 0.0)
            z *= -1.0;

        R.col(0) = x;
        R.col(1) = y;
        R.col(2) = z;
        iso.linear() = R;

        return fromIso(iso, in.header.frame_id);
    }

} // namespace behav3d::target_builder

File: src/behav3d_utils/behav3d_utils/target_builder.py
#!/usr/bin/env python3
from __future__ import annotations

from typing import Sequence, Tuple

import numpy as np
from scipy.spatial.transform import Rotation, Slerp

from builtin_interfaces.msg import Time
from geometry_msgs.msg import PoseStamped
from rclpy.clock import Clock


def _now_msg() -> Time:
    try:
        return Clock().now().to_msg()
    except Exception:
        return Time()


def pose_stamped(
    x: float,
    y: float,
    z: float,
    qx: float,
    qy: float,
    qz: float,
    qw: float,
    frame: str,
) -> PoseStamped:
    ps = PoseStamped()
    ps.header.frame_id = frame
    ps.header.stamp = _now_msg()
    ps.pose.position.x = float(x)
    ps.pose.position.y = float(y)
    ps.pose.position.z = float(z)
    ps.pose.orientation.x = float(qx)
    ps.pose.orientation.y = float(qy)
    ps.pose.orientation.z = float(qz)
    ps.pose.orientation.w = float(qw)
    return ps


def world_xy(x: float, y: float, z: float, frame: str) -> PoseStamped:
    return pose_stamped(x, y, z, 0.0, 0.0, 0.0, 1.0, frame)


def world_xz(x: float, z: float, y: float, frame: str) -> PoseStamped:
    return pose_stamped(
        x,
        y,
        z,
        0.7071067811865475,
        0.0,
        0.0,
        0.7071067811865475,
        frame,
    )


def world_yz(y: float, z: float, x: float, frame: str) -> PoseStamped:
    return pose_stamped(x, y, z, 0.5, 0.5, 0.5, 0.5, frame)


def to_iso(pose: PoseStamped) -> np.ndarray:
    mat = np.eye(4, dtype=float)
    mat[:3, :3] = Rotation.from_quat(
        [
            pose.pose.orientation.x,
            pose.pose.orientation.y,
            pose.pose.orientation.z,
            pose.pose.orientation.w,
        ]
    ).as_matrix()
    mat[:3, 3] = [
        pose.pose.position.x,
        pose.pose.position.y,
        pose.pose.position.z,
    ]
    return mat


def from_iso(mat: np.ndarray, frame: str) -> PoseStamped:
    ps = PoseStamped()
    ps.header.frame_id = frame
    ps.header.stamp = _now_msg()
    ps.pose.position.x = float(mat[0, 3])
    ps.pose.position.y = float(mat[1, 3])
    ps.pose.position.z = float(mat[2, 3])
    q = Rotation.from_matrix(mat[:3, :3]).as_quat()
    ps.pose.orientation.x = float(q[0])
    ps.pose.orientation.y = float(q[1])
    ps.pose.orientation.z = float(q[2])
    ps.pose.orientation.w = float(q[3])
    return ps


def translate(pose: PoseStamped, d: Sequence[float]) -> PoseStamped:
    mat = to_iso(pose)
    vec = np.asarray(d, dtype=float).reshape(3)
    mat[:3, 3] = mat[:3, 3] + mat[:3, :3].dot(vec)
    return from_iso(mat, pose.header.frame_id)


def rotate_euler(pose: PoseStamped, rpy: Sequence[float], degrees: bool) -> PoseStamped:
    mat = to_iso(pose)
    rot = Rotation.from_matrix(mat[:3, :3])
    delta = Rotation.from_euler("xyz", rpy, degrees=degrees)
    mat[:3, :3] = (rot * delta).as_matrix()
    return from_iso(mat, pose.header.frame_id)


def transform_rel(
    pose: PoseStamped,
    t: Sequence[float],
    q: Sequence[float],
) -> PoseStamped:
    mat = to_iso(pose)
    vec = np.asarray(t, dtype=float).reshape(3)
    mat[:3, 3] = mat[:3, 3] + mat[:3, :3].dot(vec)
    if isinstance(q, Rotation):
        delta = q
    else:
        delta = Rotation.from_quat([q[0], q[1], q[2], q[3]])
    mat[:3, :3] = (Rotation.from_matrix(mat[:3, :3]) * delta).as_matrix()
    return from_iso(mat, pose.header.frame_id)


def change_basis(basis: PoseStamped, local: PoseStamped) -> PoseStamped:
    iso_basis = to_iso(basis)
    iso_local = to_iso(local)
    iso_world = iso_basis @ iso_local
    return from_iso(iso_world, basis.header.frame_id)


def rebase(pose: PoseStamped, src: PoseStamped, dst: PoseStamped) -> PoseStamped:
    iso_src = to_iso(src)
    iso_pose = to_iso(pose)
    iso_world = iso_src @ iso_pose
    iso_dst = to_iso(dst)
    iso_new = np.linalg.inv(iso_dst) @ iso_world
    return from_iso(iso_new, dst.header.frame_id)


def _axis(pose: PoseStamped, idx: int) -> np.ndarray:
    mat = Rotation.from_quat(
        [
            pose.pose.orientation.x,
            pose.pose.orientation.y,
            pose.pose.orientation.z,
            pose.pose.orientation.w,
        ]
    ).as_matrix()
    return mat[:, idx]


def x_axis(pose: PoseStamped) -> np.ndarray:
    return _axis(pose, 0)


def y_axis(pose: PoseStamped) -> np.ndarray:
    return _axis(pose, 1)


def z_axis(pose: PoseStamped) -> np.ndarray:
    return _axis(pose, 2)


def align_target(pose: PoseStamped, desired_x: Sequence[float]) -> PoseStamped:
    mat = to_iso(pose)
    rot = mat[:3, :3]
    z = rot[:, 2]
    d = np.asarray(desired_x, dtype=float).reshape(3)

    x_proj = d - d.dot(z) * z
    if float(np.dot(x_proj, x_proj)) < 1e-10:
        fallback = rot[:, 1]
        x_proj = fallback - fallback.dot(z) * z

    x = x_proj / max(np.linalg.norm(x_proj), 1e-12)
    y = np.cross(z, x)
    y /= max(np.linalg.norm(y), 1e-12)

    rot[:, 0] = x
    rot[:, 1] = y
    rot[:, 2] = z
    mat[:3, :3] = rot
    return from_iso(mat, pose.header.frame_id)


def adjust_target(pose: PoseStamped, new_normal: Sequence[float]) -> PoseStamped:
    mat = to_iso(pose)
    rot = mat[:3, :3]
    z = np.asarray(new_normal, dtype=float).reshape(3)
    z /= max(np.linalg.norm(z), 1e-12)
    x = rot[:, 0]

    if abs(float(x.dot(z))) > 0.95:
        x = np.array([1.0, 0.0, 0.0]) if abs(float(z.dot([1.0, 0.0, 0.0]))) < 0.8 else np.array([0.0, 1.0, 0.0])

    y = np.cross(z, x)
    y /= max(np.linalg.norm(y), 1e-12)
    x = np.cross(y, z)

    rot[:, 0] = x
    rot[:, 1] = y
    rot[:, 2] = z
    mat[:3, :3] = rot
    return from_iso(mat, pose.header.frame_id)


def axis_angle(pose: PoseStamped) -> Tuple[np.ndarray, float]:
    rotvec = Rotation.from_quat(
        [
            pose.pose.orientation.x,
            pose.pose.orientation.y,
            pose.pose.orientation.z,
            pose.pose.orientation.w,
        ]
    ).as_rotvec()
    angle = float(np.linalg.norm(rotvec))
    if angle < 1e-12:
        return np.array([1.0, 0.0, 0.0]), 0.0
    return rotvec / angle, angle


def pose_between(a: PoseStamped, b: PoseStamped, t: float) -> PoseStamped:
    mat_a = to_iso(a)
    mat_b = to_iso(b)
    p = mat_a[:3, 3] * (1.0 - t) + mat_b[:3, 3] * t

    rot_a = Rotation.from_matrix(mat_a[:3, :3])
    rot_b = Rotation.from_matrix(mat_b[:3, :3])
    slerp = Slerp(
        [0.0, 1.0],
        Rotation.from_quat(np.stack([rot_a.as_quat(), rot_b.as_quat()], axis=0)),
    )
    rot = slerp([t])[0]

    mat = np.eye(4, dtype=float)
    mat[:3, :3] = rot.as_matrix()
    mat[:3, 3] = p
    return from_iso(mat, a.header.frame_id)


def mirror_across_plane(
    pose: PoseStamped,
    n: Sequence[float],
    p0: Sequence[float],
) -> PoseStamped:
    n_n = np.asarray(n, dtype=float).reshape(3)
    n_n /= max(np.linalg.norm(n_n), 1e-12)
    pos = np.array(
        [pose.pose.position.x, pose.pose.position.y, pose.pose.position.z],
        dtype=float,
    )
    p0_v = np.asarray(p0, dtype=float).reshape(3)
    mirrored_pos = pos - 2.0 * (pos - p0_v).dot(n_n) * n_n

    q = Rotation.from_quat(
        [
            pose.pose.orientation.x,
            pose.pose.orientation.y,
            pose.pose.orientation.z,
            pose.pose.orientation.w,
        ]
    )
    reflect = Rotation.from_rotvec(np.pi * n_n)
    mirrored_q = reflect * q

    return pose_stamped(
        mirrored_pos[0],
        mirrored_pos[1],
        mirrored_pos[2],
        *mirrored_q.as_quat(),
        pose.header.frame_id,
    )


def average(poses: Sequence[PoseStamped]) -> PoseStamped:
    if not poses:
        return PoseStamped()

    c = np.zeros(3, dtype=float)
    accum = np.zeros(4, dtype=float)

    for p in poses:
        c += np.array([p.pose.position.x, p.pose.position.y, p.pose.position.z], dtype=float)
        q = np.array(
            [
                p.pose.orientation.x,
                p.pose.orientation.y,
                p.pose.orientation.z,
                p.pose.orientation.w,
            ],
            dtype=float,
        )
        if q[3] < 0:
            q = -q
        accum += q

    c /= float(len(poses))
    norm = np.linalg.norm(accum)
    if norm < 1e-12:
        accum = np.array([0.0, 0.0, 0.0, 1.0], dtype=float)
    else:
        accum /= norm

    return pose_stamped(c[0], c[1], c[2], accum[0], accum[1], accum[2], accum[3], poses[0].header.frame_id)


def offset_target(pose: PoseStamped, offset_dist: float) -> PoseStamped:
    return translate(pose, [0.0, 0.0, float(offset_dist)])


def set_target_origin(pose: PoseStamped, x: float, y: float, z: float) -> PoseStamped:
    out = PoseStamped()
    out.header = pose.header
    out.header.stamp = _now_msg()
    out.pose = pose.pose
    out.pose.position.x = float(x)
    out.pose.position.y = float(y)
    out.pose.position.z = float(z)
    return out


def flip_target_axes(pose: PoseStamped, flip_x: bool, flip_y: bool) -> PoseStamped:
    mat = to_iso(pose)
    rot = mat[:3, :3]

    if flip_x:
        rot[:, 0] *= -1.0
    if flip_y:
        rot[:, 1] *= -1.0

    x = rot[:, 0] / max(np.linalg.norm(rot[:, 0]), 1e-12)
    y = rot[:, 1] / max(np.linalg.norm(rot[:, 1]), 1e-12)
    z = np.cross(x, y)
    z /= max(np.linalg.norm(z), 1e-12)
    if float(np.dot(np.cross(x, y), z)) < 0.0:
        z *= -1.0

    rot[:, 0] = x
    rot[:, 1] = y
    rot[:, 2] = z
    mat[:3, :3] = rot
    return from_iso(mat, pose.header.frame_id)


def swap_target_axes(pose: PoseStamped) -> PoseStamped:
    mat = to_iso(pose)
    rot = mat[:3, :3]
    rot[:, [0, 1]] = rot[:, [1, 0]]

    x = rot[:, 0] / max(np.linalg.norm(rot[:, 0]), 1e-12)
    y = rot[:, 1] / max(np.linalg.norm(rot[:, 1]), 1e-12)
    z = np.cross(x, y)
    z /= max(np.linalg.norm(z), 1e-12)
    if float(np.dot(np.cross(x, y), z)) < 0.0:
        z *= -1.0

    rot[:, 0] = x
    rot[:, 1] = y
    rot[:, 2] = z
    mat[:3, :3] = rot
    return from_iso(mat, pose.header.frame_id)


# Compatibility aliases for camelCase usage (C++ parity)
poseStamped = pose_stamped
worldXY = world_xy
worldXZ = world_xz
worldYZ = world_yz
toIso = to_iso
fromIso = from_iso
rotateEuler = rotate_euler
transformRel = transform_rel
changeBasis = change_basis
rebase = rebase
xAxis = x_axis
yAxis = y_axis
zAxis = z_axis
alignTarget = align_target
adjustTarget = adjust_target
axisAngle = axis_angle
poseBetween = pose_between
mirrorAcrossPlane = mirror_across_plane
offsetTarget = offset_target
setTargetOrigin = set_target_origin
flipTargetAxes = flip_target_axes
swapTargetAxes = swap_target_axes
